### PR TITLE
Draft Fix for PBC Examples

### PR DESCRIPTION
**Pull request by RylieWeaver**
@allaffa

### CODE REVIEW

File: examples/alexandria/train.py [hunk 1]
```diff
@@ -156,20 +156,23 @@ def get_magmoms_array_from_structure(structure):
...
 
         pbc = None
         try:
             pbc = structure["lattice"]["pbc"]
         except:
             print(f"Structure {entry_id} does not have pbc", flush=True)
-            return data_object
+
+        # If either cell or pbc were not read, we set to defaults which are not none.
+        if cell is None or pbc is None:
```
Comment (RylieWeaver) [inline examples/alexandria/train.py:173]: @allaffa 

If we aren't able to extract the pbc or cell from a sample where we'd expect it, would you rather set to no pbc or throw the sample out?

File: examples/alexandria/train.py [hunk 2]
```diff
@@ -156,20 +156,23 @@ def get_magmoms_array_from_structure(structure):
...
+            cell = torch.eye(3, dtype=torch.float32)
+            pbc = [False, False, False]
```
Comment (RylieWeaver) [inline examples/alexandria/train.py:175]: Does ADIOS expect the same data type for each data object as well, in addition to having the same object? In other words, if one sample has a pbc boolean and one has a pbc list of booleans, is that a problem?

File: examples/alexandria/train.py [hunk 3]
```diff
@@ -295,7 +298,7 @@ def get_magmoms_array_from_structure(structure):
         else:
             data_object.y = data_object.energy
 
-        if data_object.pbc is not None and data_object.cell is not None:
+        if any(data_object["pbc"]):
```
Comment (RylieWeaver) [inline examples/alexandria/train.py:301]: @allaffa 

A check such as this would require that data_object.pbc is of list type, which is likely not true for all datasets. Given this, and the fact that the default is set as a list of booleans, I think the best solution may be standardizing the pbc we read. So, instead of 
`pbc = structure["lattice"]["pbc"]`
we have...
`pbc = pbc_as_list(structure["lattice"]["pbc"])`

I would include this function in the utils. Let me know if you agree.

File: examples/alexandria/train.py (hunk 2 above)
Comment (RylieWeaver) [reply examples/alexandria/train.py:175]: @jychoi-hpc

Hey Jong, this may be a question well-suited to check with you, because I know you have taken part in various things related ADIOS.

File: examples/alexandria/train.py (hunk 1 above)
Comment (allaffa) [reply examples/alexandria/train.py:173]: @RylieWeaver 
I would prefer that we do NOT apply PBC and that we return an output message to declare this.

File: examples/alexandria/train.py (hunk 2 above)
Comment (allaffa) [reply examples/alexandria/train.py:175]: @kshitij-v-mehta  
Can help too

Comment (RylieWeaver) [reply examples/alexandria/train.py:175]: @kshitij-v-mehta

This may be a good question for you as well.

File: examples/alexandria/train.py (hunk 3 above)
Comment (allaffa) [reply examples/alexandria/train.py:301]: @RylieWeaver I agree

File: examples/alexandria/train.py (hunk 2 above)
Comment (kshitij-v-mehta) [reply examples/alexandria/train.py:175]: Yes, every data object must of the same type. However, we can write out a single value as a list with a single element.

I assume you want to store pbc in the ADIOS output as well? Currently, the ADIOS writer expects keys in the Data class to be of type Tensor, ndArray, floatingpoint, or integer. If you want to store a list of boolean values, let me know.

Comment (RylieWeaver) [reply examples/alexandria/train.py:175]: In principle, I believe it is good to include pbc in the data_object, even if for no reason other than completeness and debugging.

Just to double check, ADIOS we'd expect ADIOS to throw an error if one data_object.cell is an np.array and another data_object.cell is a torch.tensor, or any other type of mismatch?

How much overhead is required if we want to include a list of bools? I think we could also go with a Tensor of booleans; would that already be included in the ADIOS supported types?

Comment (allaffa) [reply examples/alexandria/train.py:175]: @RylieWeaver 

I want all data obejcts to have data.pbc stored as a torch tensor of boolean variables. 

This solves all the inconsistencies and forces a standardized choice for every user.

Comment (allaffa) [reply examples/alexandria/train.py:175]: @RylieWeaver @kshitij-v-mehta 

Since @pzanghit and I do not need PBC for tye first runs anyway, let's go with option 1. I would prefer to keep this variable as Boolean

Comment (RylieWeaver) [reply examples/alexandria/train.py:175]: It will all be the same size since I will normalize whatever pbc value is passed, whether a single boolean or 3.

File: hydragnn/preprocess/graph_samples_checks_and_updates.py [hunk 4]
```diff
@@ -285,6 +284,99 @@ def __repr__(self) -> str:
         return f"{self.__class__.__name__}(r={self.r})"
 
 
+@functional_transform("pbc_distance")
```
Comment (RylieWeaver) [inline hydragnn/preprocess/graph_samples_checks_and_updates.py:287]: I've created transforms for pbc that inherit from the Native PyTorch Geometric transforms. This way, we can guarantee the same sizing, type, and function (accounting for edge shifts) between pbc and non-pbc.

File: hydragnn/utils/datasets/adiosdataset.py [hunk 5]
```diff
@@ -175,7 +175,11 @@ def save(self):
                 arr_list = list()
                 for data in self.dataset[label]:
                     if isinstance(data[k], torch.Tensor):
-                        arr_list.append(data[k].cpu().numpy())
+                        if data[k].dtype == torch.bool:
```
Comment (RylieWeaver) [inline hydragnn/utils/datasets/adiosdataset.py:178]: @kshitij-v-mehta 

Please let me know with a if you think this is the correct way to handle tensor_bool --> int in dataset writing, and int --> tensor_bool in reading for ADIOS.

File: examples/open_catalyst_2020/utils/atoms_to_graphs.py [hunk 6]
```diff
@@ -132,19 +145,20 @@ def convert(
         else:
             data_object.y = data_object.energy
 
-        if data_object.pbc is not None and data_object.cell is not None:
+        if data_object.pbc.any():
             try:
                 data_object = self.radius_graph_pbc(data_object)
+                data_object = transform_coordinates_pbc(data_object)
```
Comment (RylieWeaver) [inline examples/open_catalyst_2020/utils/atoms_to_graphs.py:151]: @allaffa 

This transform will operate in the same way for pbc and non-pbc examples that have edge_shifts=[0, 0, 0]. However, since it is newly developed, I have left in the original transformations as a stable implementation.

File: examples/alexandria/train.py (hunk 1 above)
Comment (RylieWeaver) [reply examples/alexandria/train.py:178]: @allaffa 

If we get the either of the messages:
(1) print(f"Structure {entry_id} does not have cell", flush=True)
(2) print(f"Structure {entry_id} does not have pbc", flush=True)

We will not be applying pbc. Would you like us to have another message when the default behavior is activated, or this is enough?

File: examples/qm7x/train.py [hunk 7]
```diff
@@ -170,6 +170,9 @@ def hdf5_to_graph(self, fMOL, molid):
             pos = torch.from_numpy(np.array(fMOL[molid][confid]["atXYZ"])).to(
                 torch.float32
             )
+            cell = torch.eye(3, dtype=torch.float32)
+            pbc = torch.tensor([False, False, False], dtype=torch.bool)
+            edge_shifts = torch.tensor([0.0, 0.0, 0.0], dtype=torch.float32)
```
Comment (RylieWeaver) [inline examples/qm7x/train.py:175]: @allaffa 

As of now, I have not adjusted any of the organic molecule examples to write pbc, cell, or edge_shifts. Can you confirm that we are indeed staying without for this GFM iteration of specifically organic molecules?

File: hydragnn/utils/datasets/adiosdataset.py (hunk 5 above)
Comment (kshitij-v-mehta) [reply hydragnn/utils/datasets/adiosdataset.py:178]: Looks right to me.

File: examples/alexandria/train.py (hunk 1 above)
Comment (allaffa) [reply examples/alexandria/train.py:177]: @RylieWeaver 
I think this suffices

File: examples/omat24/train.py [hunk 8]
```diff
@@ -159,13 +165,18 @@ def __init__(
 
                     pbc = None
                     try:
-                        pbc = dataset.get_atoms(index).get_pbc()
+                        pbc = pbc_as_tensor(dataset.get_atoms(index).get_pbc())
```
Comment (allaffa) [inline examples/omat24/train.py:168]: @RylieWeaver 
Where are `get_atoms()` and `get_pbc()` defined

File: examples/open_catalyst_2020/utils/atoms_to_graphs.py [hunk 9]
```diff
@@ -97,9 +102,14 @@ def convert(
 
         pbc = None
         try:
-            pbc = atoms.get_pbc()
+            pbc = pbc_as_tensor(atoms.get_pbc())
```
Comment (allaffa) [inline examples/open_catalyst_2020/utils/atoms_to_graphs.py:105]: @RylieWeaver 
Where is ``get_pbc()` defined?

File: examples/open_catalyst_2022/train.py [hunk 10]
```diff
@@ -150,9 +155,14 @@ def ase_to_torch_geom(self, atoms):
 
         pbc = None
         try:
-            pbc = atoms.get_pbc()
+            pbc = pbc_as_tensor(atoms.get_pbc())
```
Comment (allaffa) [inline examples/open_catalyst_2022/train.py:158]: @RylieWeaver 
Where is `get_pbc()` defined?

**Review by allaffa — COMMENTED** [review]
@RylieWeaver 

Where are `get_atoms()` and `get_pbc()` defined

File: examples/omat24/train.py (hunk 8 above)
Comment (RylieWeaver) [reply examples/omat24/train.py:168]: @allaffa

All of these are functions of ASE. 

Since dataset is an ASEDBDataset, it has the `get_atoms(index)` method which yields an ASE atoms object. Then, each ASE atoms object has the `get_pbc()` and `get_cell()` methods.

https://wiki.fysik.dtu.dk/ase/_modules/ase/atoms.html#Atoms
